File: part2_section3/memory.cpp
// C++: dynamic allocation showing manual memory management
#include <iostream>
using namespace std;

int main() {
    const int SIZE = 1000000;

    // Manually allocate an array on the heap with 'new'.
    int* numbers = new int[SIZE];
    for (int i = 0; i < SIZE; i++) {
        numbers[i] = i + 1;
    }

    long long sum = 0;
    for (int i = 0; i < SIZE; i++) {
        sum += numbers[i];
    }
    cout << "C++: allocated an array of " << SIZE << " integers" << endl;
    cout << "C++: sum = " << sum << endl;

    // The programmer MUST free the memory manually.
    delete[] numbers;
    numbers = nullptr;   // avoid a dangling pointer by resetting to null
    cout << "C++: memory released manually with delete[]" << endl;
    cout << "C++: forgetting delete[] would cause a memory leak" << endl;
    return 0;
}

File: part2_section3/memory.rs
// Rust: dynamic allocation showing ownership and borrowing
fn sum_vector(data: &Vec<i64>) -> i64 {   // borrows the vector (does not take ownership)
    let mut total: i64 = 0;
    for value in data {
        total += value;
    }
    total
}

fn main() {
    // Allocate a large vector on the heap.
    let numbers: Vec<i64> = (1..=1_000_000).collect();
    println!("Rust: allocated a vector of {} integers", numbers.len());

    // Borrow the vector; ownership stays with `numbers`.
    let total = sum_vector(&numbers);
    println!("Rust: sum = {}", total);

    // `numbers` is still usable here because it was only borrowed.
    println!("Rust: first element is still accessible = {}", numbers[0]);

    // When `numbers` goes out of scope at the end of main, Rust
    // automatically frees the heap memory. No manual free, no garbage collector.
    println!("Rust: memory is released automatically when the owner goes out of scope");
}
